File: Hash.java
// Java hash table implementation
// O(1) Average complexity for search, insertion and delete

// Notes:
// supports collection size of 1k
// does not currently support chaining

public class Hash {

	//array to store hash values
	private String[] hashArray = new String[1000];

	//main method
	public static void main(String[] args) {

		//Create object
		Hash hash = new Hash();
		hash.hashingFunction("hello");
		hash.insert("hello");
		System.out.println(hash.findValue("hello"));
		
	}

	//method to return unique key
	public int hashingFunction(String key) {
		int value = 0;

		//Iterate through each character in the string
		for (char c : key.toCharArray())
		 {
		 	//get ascii value of character
		 	int ascii = (int)c;
		 	//add ascii value to the total value
		 	value += ascii;
		}

		//keep within 1k range
		value = value % 1000;

		//return the value
		return value;
	}

	// find a value in the hash table
	public String findValue(String s) {
		//perform hash on the string to determine location
		int location = hashingFunction(s);
		//get value at location
		String value = hashArray[location];

		if (value == null) {
			return "value not found";
		}
		else {
			//return value
			return value;
		}

	}

	//method to insert value into hash table
	public void insert(String s) {
		//get the hash value for the string
		int value = hashingFunction(s);
		//store the value in the hash table
		hashArray[value] = s;
	}

	//method to delete value from hash table
	public void delete(String s) {
		//get the hash value for the string
		int value = hashingFunction(s);
		//store the value in the hash table
		hashArray[value] = null;
	}

	

}

File: hash.rb
require 'pry'

# Hash table implementation
# O(1) Average complexity for search, insertion and delete

# Notes:
# supports collection size of 1k
# does not currently support chaining

class HashTable

  # hashed collection
  @@collection = Array.new

  # Hashing function
  def hash_function(value)
    # total
    total = 0

    # iterate over each character
    value.each_char do |char|
      # get the ascii value
      ascii = char.ord
      # add value of character to total
      total += ascii
      # modulo 1000 (table range)
      total = total % 1000
    end

    # return result
    total

  end

  # insert value into collection
  def insert(value)
    #get hashed value
    hash_value = hash_function(value)
    # insert value
    @@collection[hash_value] = value

    # return confirmation
    "value: " << value << " inserted at position: " << hash_value.to_s
  end

  # delete value from collection
  def delete(value)
    #get hashed value
    hash_value = hash_function(value)

    # check if value exists
    result = search(value)

    # if value exists
    if result
      # set value to nil
      @@collection[hash_value] = nil
      # return confirmation
      "value deleted"
    else
      # return error message
      "value not found"
    end


  end

  # search collection for value
  def search(value)
    #get hashed value
    hash_value = hash_function(value)

    # if value exists return true
    if @@collection[hash_value] == value
      return true
    else
      return false
    end
  end

  # function to test various hash table features
  def self.test
    # tests
    h = HashTable.new
    # test hash function
    puts h.hash_function("test")
    # test insert
    puts h.insert("test")
    # test search
    puts h.search("test")
    # test delete
    puts h.delete("test")

  end

end

# run test
HashTable.test
